service/matching/workers/registry_impl.go

```diff
@@ -0,0 +1,278 @@
+	now := time.Now()
+	key := hb.WorkerInstanceKey
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	mp, ok := b.namespaces[nsID]
+	if !ok {
+		mp = make(map[string]*entry)
+		b.namespaces[nsID] = mp
+	}
+
+	if e, exists := mp[key]; exists {
+		e.hb = hb
+		e.lastSeen = now
+		b.order.MoveToBack(e.elem)
+		return false
+	}
+
+	e := &entry{
+		nsID:     nsID,
+		hb:       hb,
+		lastSeen: now,
+	}
+	e.elem = b.order.PushBack(e)
+	mp[key] = e
+	return true
+}
+
+// evictByTTL removes entries older than expireBefore from this bucket.
+// Returns the number of entries removed.
+func (b *bucket) evictByTTL(expireBefore time.Time) int {
+	removed := 0
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	for {
+		front := b.order.Front()
+		if front == nil {
+			break
+		}
+		e := front.Value.(*entry) //nolint:revive
+		if !e.lastSeen.Before(expireBefore) {
+			break
+		}
+		b.order.Remove(front)
+		delete(b.namespaces[e.nsID], e.hb.WorkerInstanceKey)
+		removed++
+	}
+	return removed
+}
+
+// NewRegistry creates a workers heartbeat registry with the given parameters.
+func NewRegistry(lc fx.Lifecycle) Registry {
+	m := newRegistryImpl(
+		defaultBuckets,
+		defaultEntryTTL,
+		defaultMinEvictAge,
+		defaultMaxEntries,
+		defaultEvictionInterval,
+	)
+
+	lc.Append(fx.Hook{
+		OnStart: func(ctx context.Context) error {
+			m.Start()
+			return nil
+		},
+		OnStop: func(ctx context.Context) error {
+			m.Stop()
+			return nil
+		},
+	})
+
+	return m
+}
+
+func newRegistryImpl(numBuckets int,
+	ttl time.Duration,
+	minEvictAge time.Duration,
+	maxItems int64,
+	evictionInterval time.Duration,
+) *registryImpl {
+	m := &registryImpl{
+		buckets:          make([]*bucket, numBuckets),
+		maxItems:         maxItems,
+		tTL:              ttl,
+		minEvictAge:      minEvictAge,
+		evictionInterval: evictionInterval,
+		hasher:           &maphash.Hash{},
+		quit:             make(chan struct{}),
+	}
+	m.hasher.Reset()
+
+	for i := range m.buckets {
+		m.buckets[i] = newBucket()
+	}
+	return m
+}
+
+// bucketFor hashes the namespace to select a bucket.
+func (m *registryImpl) getBucket(nsID namespace.ID) *bucket {
+	m.hasher.Reset()
+	m.hasher.WriteString(nsID.String()) //nolint:revive
+	hs := m.hasher.Sum64()
+	idx := int(hs % uint64(len(m.buckets)))
+
+	return m.buckets[idx]
+}
+
+// upsertHeartbeat records or refreshes a WorkerHeartbeat under the given namespace.
+// New entries increment the global counter.
+func (m *registryImpl) upsertHeartbeat(nsID namespace.ID, hb *workerpb.WorkerHeartbeat) {
+	b := m.getBucket(nsID)
+	if newEntry := b.upsertHeartbeat(nsID, hb); newEntry {
+		m.total.Add(1)
+	}
+}
+
+// listWorkers returns all WorkerHeartbeats in a namespace
+// for which predicate(hb) returns true.
+func (m *registryImpl) filterWorkers(
+	nsID namespace.ID,
+	predicate func(*workerpb.WorkerHeartbeat) bool,
+) []*workerpb.WorkerHeartbeat {
+	b := m.getBucket(nsID)
+	b.mu.Lock()
+	defer b.mu.Unlock()
```

Comment:
the body of this should probably be moved into a method on `bucket`? the registry shouldn't be locking+manipulating the bucket